sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py

```diff
@@ -567,3 +567,69 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
         assert predictor.get_max_depth() < predictor_ohe.get_max_depth()
 
     np.testing.assert_allclose(preds, preds_ohe)
+
+
+def test_grower_interaction_constraints():
+    """Check that grower respects interaction constraints."""
+    n_features = 6
+    interaction_cst = [{0, 1}, {1, 2}, {3, 4, 5}]
+    n_samples = 10
+    n_bins = 6
+    root_feature_splits = []
+
+    def get_all_children(node):
+        res = []
+        if node.is_leaf:
+            return res
+        for n in [node.left_child, node.right_child]:
+            res.append(n)
+            if not n.is_leaf:
+                res.extend(get_all_children(n))
+        return res
+
+    for seed in range(20):
+        rng = np.random.RandomState(seed)
+
+        X_binned = rng.randint(
+            0, n_bins - 1, size=(n_samples, n_features), dtype=X_BINNED_DTYPE
+        )
+        X_binned = np.asfortranarray(X_binned)
+        gradients = rng.normal(size=n_samples).astype(G_H_DTYPE)
+        hessians = np.ones(shape=1, dtype=G_H_DTYPE)
+
+        grower = TreeGrower(
+            X_binned,
+            gradients,
+            hessians,
+            n_bins=n_bins,
+            min_samples_leaf=1,
+            interaction_cst=interaction_cst,
+            n_threads=n_threads,
+        )
+        grower.grow()
+
+        root_feature_idx = grower.root.split_info.feature_idx
+        root_feature_splits.append(root_feature_idx)
+
+        feature_idx_to_constraint_set = {
+            0: {0, 1},
+            1: {0, 1, 2},
+            2: {1, 2},
+            3: {3, 4, 5},
+            4: {3, 4, 5},
+            5: {3, 4, 5},
+        }
+
+        root_constraint_set = feature_idx_to_constraint_set[root_feature_idx]
+        for node in (grower.root.left_child, grower.root.right_child):
+            # Root's children's allowed_features must be the root's constraints set.
+            assert_array_equal(node.allowed_features, list(root_constraint_set))
+        for node in get_all_children(grower.root):
+            # Nodes accessible from the root must have their index in the root's
+            # constraints set. For example, sets {0, 1} and {1, 2} must not interact
+            # with each other.
+            if not node.is_leaf:
+                assert node.split_info.feature_idx in root_constraint_set
```

Comment:
I think the comment does not match what is being tested, here.

This is rather:
```suggestion
            # Nodes accessible from the root must have their index in the root's
            # constraints set.
            if not node.is_leaf:
                assert node.split_info.feature_idx in root_constraint_set
```

The other part mentioned (i.e. "For example, sets {0, 1} and {1, 2} must not interact with each other.") must be tested (separately).

I think that the best way to test it is to recursively check that:

```python
   parent_interaction_cst_indices = set(node.interaction_cst_indices.flatten())
   right_interactions_cst_indices = set(
       node.right_child.interaction_cst_indices.flatten()
   )
   left_interactions_cst_indices = set(
       node.left_child.interaction_cst_indices.flatten()
   )

   assert parent_interaction_cst_indices.issupersset(right_interactions_cst_indices)
   assert parent_interaction_cst_indices.issupersset(left_interactions_cst_indices)
```

What do you think?

Reply:
Good catch. You're right. I adapted your version with slight changes.

Reply:
See https://github.com/scikit-learn/scikit-learn/pull/21020/commits/28c45787006fb6803fa62c61601f6ea8948c194f.